### ISSUE TITLE
logstash mutate add_field changes type of referenced field

### ISSUE THREAD

**Issue by jamesblackburn** [elastic/logstash#6774]
- Version: 5.2.1
- Operating System: Linux RHEL 7
- Config File (if you have sensitive info, please remove it):

We have some filter config which looks like this:

```
            # Split load averages into separate fields
            mutate {
                add_field => {
                    "os[load_average_1]" => "%{os[load_average][0]}"
                    "os[load_average_5]" => "%{os[load_average][1]}"
                    "os[load_average_15]" => "%{os[load_average][2]}"
                }
            }
```

Where `os[load_average][0]` is a `float` field.  

In logstash `1.5.5` the output fields `load_average_1` etc. were emitted as floats.

In logstash 5.2.1 I need to explicitly convert the fields to float like:

```
            mutate {
	            convert => { 
	            	"[os][load_average_1]" => "float"
	            	"[os][load_average_5]" => "float"
	            	"[os][load_average_15]" => "float" 
	            }
            }
```

Is this expected?

**Comment by untergeek** [comment]
I think it should be, in order to remain consistent.  

Consider this:
```
"os[load_average_1]" => "%{os[load_average][0]}"
```
You have encapsulated the value of `os[load_average][0]` within double-quotes.  That makes it a string.  Unless you want to use the ruby filter, there is no other way to pass the value of one field to another _within Logstash._ 

Now, if you're sending this to Elasticsearch, it does not have to be a float value in the output.  The Elasticsearch mapping can do the conversion for you.  You would have to create that in a mapping template, or manually put the mapping in place _before_ you send the data through Logstash.

**Comment by jamesblackburn** [comment]
Ok, thanks. I was surprised by the behaviour change between versions...

Feel free to close.

**Comment by jordansissel** [comment]
I think the behavior you saw was accidental. The current behavior (it's all strings for add_field) is expected.

For your situation, based on how I read your example, you could consider using `rename` instead of trying to use `add_field` and `convert`. 

Alternately, we could consider adding a new way to copy parts of an event around, if it is the case that you don't want to rename the field.

**Comment by untergeek** [comment]
@jamesblackburn Just to put your mind at ease, what you experienced was in fact happening, but it was not supposed to.  You were enjoying the benefits of a bug.  We fixed the bug in Logstash 2.2.0 when we switched to having the core event logic be in Java.
